arbnode/espresso_caff_node.go

```diff
@@ -376,27 +452,90 @@ func (n *EspressoCaffNode) Start(ctx context.Context) error {
+		blockhash, err := getHashOverInterface(currentBlock)
+		if err != nil {
+			log.Error("failed to get hash of current block", "err", err)
+			return fmt.Errorf("failed to get hash of current block: %w", err)
+		}
+
+		// Get the block signature
+		blockSignature, err := getBlockSignature(n.db, currentBlock.NumberU64())
+		if err != nil {
+			return fmt.Errorf("failed to get block signature: %w", err)
+		}
+
+		publicKeyBytes, err := crypto.Ecrecover(blockhash, blockSignature)
+		if err != nil {
+			return fmt.Errorf("failed to recover public key from signature: %w", err)
+		}
+		pubKey, err := crypto.UnmarshalPubkey(publicKeyBytes)
+		if err != nil || pubKey == nil {
+			return fmt.Errorf("failed to unmarshal public key: %w", err)
+		}
+		// Public Key to address
+		publicKeyAddress := crypto.PubkeyToAddress(*pubKey)
+		// TODO: In follow up PRs, we should allows any valid PCR0 address registered in the contract
+		// to be able to decrypt the snapshot
+		if publicKeyAddress != *n.snapshotAddress {
+			return fmt.Errorf("snapshot address does not match from address")
+		}
+
+	}
+
+	// TODO: In follow up PRs, think about how to handle the case when on initial startup we dont have a signature over the
+	// from block?
+	n.currentBlock = currentBlock
+
+	currentBlockNum := currentBlockHeader.Number.Uint64() + 1
 	currentMessagePos, err := n.executionEngine.BlockNumberToMessageIndex(currentBlockNum)
 	if err != nil {
 		return fmt.Errorf("failed to convert block number to message index: %w", err)
 	}
+
 	var nextHotshotBlock uint64
+	var nextHotshotBlockSignature []byte
 
 	if !n.configFetcher().Dangerous.IgnoreDatabaseHotshotBlock {
-		nextHotshotBlock, err = n.espressoStreamer.ReadNextHotshotBlockFromDb(n.db)
+		nextHotshotBlock, nextHotshotBlockSignature, err = n.espressoStreamer.ReadNextHotshotBlockFromDb(n.db)
 		if err != nil {
 			return fmt.Errorf("failed to read next hotshot block: %w", err)
 		}
-	}
 
+		if n.configFetcher().EspressoTeeType != "" && nextHotshotBlock != 0 {
+			hotshotBlockHash, err := getHashOverUint64(nextHotshotBlock)
+			if err != nil {
+				return fmt.Errorf("failed to get hash of hotshot block: %w", err)
+			}
+
+			publicKeyBytes, err := crypto.Ecrecover(hotshotBlockHash, nextHotshotBlockSignature)
+			if err != nil {
+				return fmt.Errorf("failed to recover public key from signature for hotshot block: %w", err)
+			}
+			pubKey, err := crypto.UnmarshalPubkey(publicKeyBytes)
+			if err != nil || pubKey == nil {
+				return fmt.Errorf("failed to unmarshal public key for hotshot block: %w", err)
+			}
+			// Public Key to address
+			publicKeyAddress := crypto.PubkeyToAddress(*pubKey)
+			// TODO: In follow up PRs, we should allows any valid PCR0 address registered in the contract
+			// to be able to decrypt the snapshot
+			if publicKeyAddress != *n.snapshotAddress {
+				return fmt.Errorf("public key address does not match from address for hotshot block")
+			}
+		}
+	}
```

Comment:
It seems like we use this same code block twice. Maybe it's worth defining a function that fetches all of this info for us to clean up the local code in the already existing functions? something like
```
hotshotBlockHash, ..., publicKeyAddress = fetchSignedInfoFromDB()
```
or something similar depending on which bits of the info are needed at different parts of the code base.